// @eth-optimism/contracts/build/test/contracts/OVM/execution/OVM_ExecutionManager/ovmSTATICCALL.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../../../../helpers");
const test_ovmSTATICCALL = {
    name: 'Basic tests for ovmSTATICCALL',
    preState: {
        ExecutionManager: {
            ovmStateManager: '$OVM_STATE_MANAGER',
            ovmSafetyChecker: '$OVM_SAFETY_CHECKER',
            messageRecord: {
                nuisanceGasLeft: helpers_1.OVM_TX_GAS_LIMIT,
            },
        },
        StateManager: {
            owner: '$OVM_EXECUTION_MANAGER',
            accounts: {
                $DUMMY_OVM_ADDRESS_1: {
                    codeHash: helpers_1.NON_NULL_BYTES32,
                    ethAddress: '$OVM_CALL_HELPER',
                },
                $DUMMY_OVM_ADDRESS_2: {
                    codeHash: helpers_1.NON_NULL_BYTES32,
                    ethAddress: '$OVM_CALL_HELPER',
                },
                $DUMMY_OVM_ADDRESS_3: {
                    codeHash: helpers_1.NON_NULL_BYTES32,
                    ethAddress: '$OVM_CALL_HELPER',
                },
            },
            contractStorage: {
                $DUMMY_OVM_ADDRESS_1: {
                    [helpers_1.NON_NULL_BYTES32]: helpers_1.getStorageXOR(helpers_1.NULL_BYTES32),
                },
                $DUMMY_OVM_ADDRESS_3: {
                    [helpers_1.NON_NULL_BYTES32]: helpers_1.getStorageXOR(helpers_1.NULL_BYTES32),
                },
            },
            verifiedContractStorage: {
                $DUMMY_OVM_ADDRESS_1: {
                    [helpers_1.NON_NULL_BYTES32]: true,
                },
                $DUMMY_OVM_ADDRESS_3: {
                    [helpers_1.NON_NULL_BYTES32]: true,
                },
            },
        },
    },
    parameters: [
        {
            name: 'ovmSTATICCALL => ovmSSTORE',
            steps: [
                {
                    functionName: 'ovmSTATICCALL',
                    functionParams: {
                        gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                        target: '$DUMMY_OVM_ADDRESS_1',
                        subSteps: [
                            {
                                functionName: 'ovmSSTORE',
                                functionParams: {
                                    key: helpers_1.NULL_BYTES32,
                                    value: helpers_1.NULL_BYTES32,
                                },
                                expectedReturnStatus: false,
                                expectedReturnValue: {
                                    flag: helpers_1.REVERT_FLAGS.STATIC_VIOLATION,
                                    nuisanceGasLeft: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                },
                            },
                        ],
                    },
                    expectedReturnStatus: false,
                },
            ],
        },
        {
            name: 'ovmSTATICCALL => ovmSLOAD',
            steps: [
                {
                    functionName: 'ovmSTATICCALL',
                    functionParams: {
                        gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                        target: '$DUMMY_OVM_ADDRESS_1',
                        subSteps: [
                            {
                                functionName: 'ovmSLOAD',
                                functionParams: {
                                    key: helpers_1.NON_NULL_BYTES32,
                                },
                                expectedReturnStatus: true,
                                expectedReturnValue: helpers_1.NULL_BYTES32,
                            },
                        ],
                    },
                    expectedReturnStatus: true,
                },
            ],
        },
        {
            name: 'ovmSTATICCALL => ovmCREATE',
            steps: [
                {
                    functionName: 'ovmSTATICCALL',
                    functionParams: {
                        gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                        target: '$DUMMY_OVM_ADDRESS_1',
                        subSteps: [
                            {
                                functionName: 'ovmCREATE',
                                functionParams: {
                                    bytecode: helpers_1.DUMMY_BYTECODE,
                                },
                                expectedReturnStatus: false,
                                expectedReturnValue: {
                                    flag: helpers_1.REVERT_FLAGS.STATIC_VIOLATION,
                                    nuisanceGasLeft: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                },
                            },
                        ],
                    },
                    expectedReturnStatus: false,
                },
            ],
        },
        {
            name: 'ovmCALL(ADDRESS_1) => ovmSTATICCALL(ADDRESS_2) => ovmCALLER',
            steps: [
                {
                    functionName: 'ovmCALL',
                    functionParams: {
                        gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                        target: '$DUMMY_OVM_ADDRESS_1',
                        subSteps: [
                            {
                                functionName: 'ovmSTATICCALL',
                                functionParams: {
                                    gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                    target: '$DUMMY_OVM_ADDRESS_2',
                                    subSteps: [
                                        {
                                            functionName: 'ovmCALLER',
                                            expectedReturnValue: '$DUMMY_OVM_ADDRESS_1',
                                        },
                                    ],
                                },
                                expectedReturnStatus: true,
                            },
                        ],
                    },
                    expectedReturnStatus: true,
                },
            ],
        },
        {
            name: 'ovmCALL(ADDRESS_1) => ovmSTATICCALL(ADDRESS_2) => ovmADDRESS',
            steps: [
                {
                    functionName: 'ovmCALL',
                    functionParams: {
                        gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                        target: '$DUMMY_OVM_ADDRESS_1',
                        subSteps: [
                            {
                                functionName: 'ovmSTATICCALL',
                                functionParams: {
                                    gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                    target: '$DUMMY_OVM_ADDRESS_2',
                                    subSteps: [
                                        {
                                            functionName: 'ovmADDRESS',
                                            expectedReturnValue: '$DUMMY_OVM_ADDRESS_2',
                                        },
                                    ],
                                },
                                expectedReturnStatus: true,
                            },
                        ],
                    },
                    expectedReturnStatus: true,
                },
            ],
        },
        {
            name: 'ovmCALL(ADDRESS_1) => ovmSTATICCALL(ADDRESS_2) => ovmCALL(ADDRESS_3) => ovmSSTORE',
            steps: [
                {
                    functionName: 'ovmCALL',
                    functionParams: {
                        gasLimit: helpers_1.OVM_TX_GAS_LIMIT,
                        target: '$DUMMY_OVM_ADDRESS_1',
                        subSteps: [
                            {
                                functionName: 'ovmSTATICCALL',
                                functionParams: {
                                    gasLimit: helpers_1.OVM_TX_GAS_LIMIT,
                                    target: '$DUMMY_OVM_ADDRESS_2',
                                    subSteps: [
                                        {
                                            functionName: 'ovmCALL',
                                            functionParams: {
                                                gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                                target: '$DUMMY_OVM_ADDRESS_3',
                                                subSteps: [
                                                    {
                                                        functionName: 'ovmSSTORE',
                                                        functionParams: {
                                                            key: helpers_1.NULL_BYTES32,
                                                            value: helpers_1.NULL_BYTES32,
                                                        },
                                                        expectedReturnStatus: false,
                                                        expectedReturnValue: {
                                                            flag: helpers_1.REVERT_FLAGS.STATIC_VIOLATION,
                                                            nuisanceGasLeft: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                                        },
                                                    },
                                                ],
                                            },
                                            expectedReturnStatus: false,
                                        },
                                    ],
                                },
                                expectedReturnStatus: true,
                            },
                        ],
                    },
                    expectedReturnStatus: true,
                },
            ],
        },
        {
            name: 'ovmCALL(ADDRESS_1) => ovmSTATICCALL(ADDRESS_2) => ovmCALL(ADDRESS_3) => ovmSLOAD',
            steps: [
                {
                    functionName: 'ovmCALL',
                    functionParams: {
                        gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                        target: '$DUMMY_OVM_ADDRESS_1',
                        subSteps: [
                            {
                                functionName: 'ovmSTATICCALL',
                                functionParams: {
                                    gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                    target: '$DUMMY_OVM_ADDRESS_2',
                                    subSteps: [
                                        {
                                            functionName: 'ovmCALL',
                                            functionParams: {
                                                gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                                target: '$DUMMY_OVM_ADDRESS_3',
                                                subSteps: [
                                                    {
                                                        functionName: 'ovmSLOAD',
                                                        functionParams: {
                                                            key: helpers_1.NON_NULL_BYTES32,
                                                        },
                                                        expectedReturnStatus: true,
                                                        expectedReturnValue: helpers_1.NULL_BYTES32,
                                                    },
                                                ],
                                            },
                                            expectedReturnStatus: true,
                                        },
                                    ],
                                },
                                expectedReturnStatus: true,
                            },
                        ],
                    },
                    expectedReturnStatus: true,
                },
            ],
        },
        {
            name: 'ovmCALL(ADDRESS_1) => ovmSTATICCALL(ADDRESS_2) => ovmCALL(ADDRESS_3) => ovmCREATE',
            steps: [
                {
                    functionName: 'ovmCALL',
                    functionParams: {
                        gasLimit: helpers_1.OVM_TX_GAS_LIMIT,
                        target: '$DUMMY_OVM_ADDRESS_1',
                        subSteps: [
                            {
                                functionName: 'ovmSTATICCALL',
                                functionParams: {
                                    gasLimit: helpers_1.OVM_TX_GAS_LIMIT,
                                    target: '$DUMMY_OVM_ADDRESS_2',
                                    subSteps: [
                                        {
                                            functionName: 'ovmCALL',
                                            functionParams: {
                                                gasLimit: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                                target: '$DUMMY_OVM_ADDRESS_3',
                                                subSteps: [
                                                    {
                                                        functionName: 'ovmCREATE',
                                                        functionParams: {
                                                            bytecode: helpers_1.DUMMY_BYTECODE,
                                                        },
                                                        expectedReturnStatus: false,
                                                        expectedReturnValue: {
                                                            flag: helpers_1.REVERT_FLAGS.STATIC_VIOLATION,
                                                            nuisanceGasLeft: helpers_1.OVM_TX_GAS_LIMIT / 2,
                                                        },
                                                    },
                                                ],
                                            },
                                            expectedReturnStatus: false,
                                        },
                                    ],
                                },
                                expectedReturnStatus: true,
                            },
                        ],
                    },
                    expectedReturnStatus: true,
                },
            ],
        },
    ],
};
const runner = new helpers_1.ExecutionManagerTestRunner();
runner.run(test_ovmSTATICCALL);
