// @eth-optimism/contracts/build/test/helpers/dummy/batches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DUMMY_BATCH_PROOFS = exports.DUMMY_BATCH_HEADERS = void 0;
const constants_1 = require("../constants");
const hardhat_1 = require("hardhat");
exports.DUMMY_BATCH_HEADERS = [
    {
        batchIndex: 0,
        batchRoot: constants_1.NULL_BYTES32,
        batchSize: 0,
        prevTotalElements: 0,
        extraData: hardhat_1.ethers.utils.defaultAbiCoder.encode(['uint256', 'address'], [constants_1.NULL_BYTES32, constants_1.NON_ZERO_ADDRESS]),
    },
    {
        batchIndex: 1,
        batchRoot: constants_1.NULL_BYTES32,
        batchSize: 0,
        prevTotalElements: 0,
        extraData: hardhat_1.ethers.utils.defaultAbiCoder.encode(['uint256', 'address'], [constants_1.NULL_BYTES32, constants_1.NON_ZERO_ADDRESS]),
    },
];
exports.DUMMY_BATCH_PROOFS = [
    {
        index: 0,
        siblings: [constants_1.NULL_BYTES32],
    },
    {
        index: 1,
        siblings: [constants_1.NULL_BYTES32],
    },
];
