// @eth-optimism/contracts/build/test/helpers/codec/revert-flags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.REVERT_FLAGS = exports.decodeRevertData = exports.encodeRevertData = void 0;
const hardhat_1 = require("hardhat");
const encodeRevertData = (flag, data = '0x', nuisanceGasLeft = 0, ovmGasRefund = 0) => {
    const abiEncoded = hardhat_1.ethers.utils.defaultAbiCoder.encode(['uint256', 'uint256', 'uint256', 'bytes'], [flag, nuisanceGasLeft, ovmGasRefund, data]);
    return abiEncoded;
};
exports.encodeRevertData = encodeRevertData;
const decodeRevertData = (revertData) => {
    const decoded = hardhat_1.ethers.utils.defaultAbiCoder.decode(['uint256', 'uint256', 'uint256', 'bytes'], revertData);
    return ('[revertFlag:' +
        Object.keys(exports.REVERT_FLAGS)[decoded[0]] +
        ', nuisanceGasLeft:' +
        decoded[1] +
        ', ovmGasRefund: ' +
        decoded[2] +
        ', data: ' +
        decoded[3] +
        ']');
};
exports.decodeRevertData = decodeRevertData;
exports.REVERT_FLAGS = {
    DID_NOT_REVERT: 0,
    OUT_OF_GAS: 1,
    INTENTIONAL_REVERT: 2,
    EXCEEDS_NUISANCE_GAS: 3,
    INVALID_STATE_ACCESS: 4,
    UNSAFE_BYTECODE: 5,
    CREATE_COLLISION: 6,
    STATIC_VIOLATION: 7,
    CREATE_EXCEPTION: 8,
    CREATOR_NOT_ALLOWED: 9,
};
