// @eth-optimism/core-utils/build/src/common/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const pino_1 = __importDefault(require("pino"));
class Logger {
    constructor(options) {
        this.options = options;
        const loggerOptions = {
            name: options.name,
            level: options.level || 'debug',
            base: null,
        };
        this.inner = options.destination
            ? pino_1.default(loggerOptions, pino_1.default.destination(options.destination))
            : pino_1.default(loggerOptions);
    }
    child(bindings) {
        const inner = this.inner.child(bindings);
        const logger = new Logger(this.options);
        logger.inner = inner;
        return logger;
    }
    trace(msg, o, ...args) {
        if (o) {
            this.inner.trace(o, msg, ...args);
        }
        else {
            this.inner.trace(msg, ...args);
        }
    }
    debug(msg, o, ...args) {
        if (o) {
            this.inner.debug(o, msg, ...args);
        }
        else {
            this.inner.debug(msg, ...args);
        }
    }
    info(msg, o, ...args) {
        if (o) {
            this.inner.info(o, msg, ...args);
        }
        else {
            this.inner.info(msg, ...args);
        }
    }
    warn(msg, o, ...args) {
        if (o) {
            this.inner.warn(o, msg, ...args);
        }
        else {
            this.inner.warn(msg, ...args);
        }
    }
    warning(msg, o, ...args) {
        if (o) {
            this.inner.warn(o, msg, ...args);
        }
        else {
            this.inner.warn(msg, ...args);
        }
    }
    error(msg, o, ...args) {
        if (o) {
            this.inner.error(o, msg, ...args);
        }
        else {
            this.inner.error(msg, ...args);
        }
    }
    fatal(msg, o, ...args) {
        if (o) {
            this.inner.fatal(o, msg, ...args);
        }
        else {
            this.inner.fatal(msg, ...args);
        }
    }
    crit(msg, o, ...args) {
        if (o) {
            this.inner.fatal(o, msg, ...args);
        }
        else {
            this.inner.fatal(msg, ...args);
        }
    }
    critical(msg, o, ...args) {
        if (o) {
            this.inner.fatal(o, msg, ...args);
        }
        else {
            this.inner.fatal(msg, ...args);
        }
    }
}
exports.Logger = Logger;
